File: bench/text/text_bench.pas
program text_bench;
{$mode objfpc}{$H+}

{
  Text Operations Micro-Benchmark: Pascal vs Go vs Rust

  Tracks:
    IntToStr/100k    — format 100000 integers to strings
    Base64Enc/4KB    — Base64 encode 4096 bytes
    Base64Dec/5.3KB  — Base64 decode 5464-char string
    HexEnc/1KB       — hex-encode 1024 bytes
    StrReplace/10KB  — StringReplace in 5.2KB string × 10000
    JSON/Parse/4KB   — parse 4KB JSON object × 10000
}

uses
  SysUtils, Classes,
  nextpas.core.base,
  nextpas.core.time.base,
  nextpas.core.bench,
  nextpas.core.bench.intf,
  nextpas.core.text.conv,
  nextpas.core.encoding,
  nextpas.core.json;

const
  INT_COUNT     = 100000;
  B64_SRC_SIZE  = 4096;
  HEX_SRC_SIZE  = 1024;
  REPLACE_N     = 10000;
  JSON_PARSE_N  = 10000;

var
  GIntResults: array[0..INT_COUNT-1] of string;
  GB64Src: TBytes;
  GB64Encoded: string;
  GHexSrc: TBytes;
  GReplaceSrc: string;
  GJsonStr: string;

procedure InitData;
var
  I: Integer;
begin
  SetLength(GB64Src, B64_SRC_SIZE);
  for I := 0 to B64_SRC_SIZE - 1 do
    GB64Src[I] := Byte(I mod 256);

  GB64Encoded := Base64Encode(GB64Src);

  SetLength(GHexSrc, HEX_SRC_SIZE);
  for I := 0 to HEX_SRC_SIZE - 1 do
    GHexSrc[I] := Byte(I mod 256);

  GReplaceSrc := '';
  for I := 1 to 100 do
    GReplaceSrc := GReplaceSrc + 'Hello World! This is a test string for replacement. ';

  GJsonStr := '{"users":[{"id":1,"name":"Alice","email":"alice@example.com","score":95.5,"active":true},' +
    '{"id":2,"name":"Bob","email":"bob@example.com","score":87.3,"active":false},' +
    '{"id":3,"name":"Charlie","email":"charlie@example.com","score":92.1,"active":true},' +
    '{"id":4,"name":"Diana","email":"diana@example.com","score":88.8,"active":true},' +
    '{"id":5,"name":"Eve","email":"eve@example.com","score":91.0,"active":false}]}';
end;

{ === IntToStr × 100000 === }

procedure BenchIntToStr(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 0 to INT_COUNT - 1 do
    GIntResults[I] := IntToStr(I);
  ACtx.SetBytes(INT_COUNT);
end;

{ === Base64 Encode 4KB === }

procedure BenchBase64Enc(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 1 to 1000 do
    Base64Encode(GB64Src);
  ACtx.SetBytes(B64_SRC_SIZE * 1000);
end;

{ === Base64 Decode 5.3KB === }

procedure BenchBase64Dec(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 1 to 1000 do
    Base64Decode(GB64Encoded);
  ACtx.SetBytes(Length(GB64Encoded) * 1000);
end;

{ === Hex Encode 1KB === }

procedure BenchHexEnc(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 1 to 1000 do
    HexEncode(GHexSrc);
  ACtx.SetBytes(HEX_SRC_SIZE * 1000);
end;

{ === StringReplace × 10000 === }

procedure BenchStrReplace(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 1 to REPLACE_N do
    StringReplace(GReplaceSrc, 'Hello', 'World', True);
  ACtx.SetBytes(REPLACE_N * Length(GReplaceSrc));
end;

{ === JSON Parse 4KB × 10000 === }

procedure BenchJsonParse(const ACtx: IBenchContext);
var
  I: Integer;
begin
  for I := 1 to JSON_PARSE_N do
    JsonParse(GJsonStr);
  ACtx.SetBytes(JSON_PARSE_N * Length(GJsonStr));
end;

{ === Main === }

var
  LSuite: IBenchSuite;
  LResults: IBenchResults;
begin
  InitData;

  WriteLn('=== nextPas Text Operations Benchmark ===');
  WriteLn('IntToStr: ', INT_COUNT, ' integers');
  WriteLn('Base64:   ', B64_SRC_SIZE, 'B encode / ', Length(GB64Encoded), 'B decode');
  WriteLn('Hex:      ', HEX_SRC_SIZE, 'B encode');
  WriteLn('Replace:  ', Length(GReplaceSrc), 'B string × ', REPLACE_N, ' iters');
  WriteLn('JSON:     ', Length(GJsonStr), 'B parse × ', JSON_PARSE_N, ' iters');
  WriteLn;

  LSuite := TBenchSuite.Create('Text')
    .SetMinDuration(TDuration.FromMilliseconds(100))
    .SetMaxIterations(10000)
    .SetMinSamples(6)
    .SetWarmupIters(3);

  LSuite.Add('IntToStr/100k', @BenchIntToStr);
  LSuite.Add('Base64Enc/4KB', @BenchBase64Enc);
  LSuite.Add('Base64Dec/5.3KB', @BenchBase64Dec);
  LSuite.Add('HexEnc/1KB', @BenchHexEnc);
  LSuite.Add('StrReplace/10KB', @BenchStrReplace);
  LSuite.Add('JSON/Parse/4KB', @BenchJsonParse);

  LResults := LSuite.Run;

  WriteLn;
  WriteLn('=== benchstat format ===');
  WriteLn(LResults.ToBenchStat);
end.

File: bench/text/benches/text_bench.rs
use criterion::{criterion_group, criterion_main, Criterion, Throughput};
use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;

fn bench_inttostr_100k(c: &mut Criterion) {
    let mut group = c.benchmark_group("IntToStr/100k");
    group.throughput(Throughput::Elements(100_000));
    group.bench_function("IntToStr/100k", |b| {
        let ints: Vec<i64> = (0..100_000).collect();
        b.iter(|| {
            let mut results = Vec::with_capacity(100_000);
            for &v in &ints {
                results.push(v.to_string());
            }
            results
        });
    });
    group.finish();
}

fn bench_base64_enc_4k(c: &mut Criterion) {
    let mut group = c.benchmark_group("Base64Enc/4KB");
    group.throughput(Throughput::Bytes(4096));
    let src: Vec<u8> = (0..4096u32).map(|i| (i % 256) as u8).collect();
    group.bench_function("Base64Enc/4KB", |b| {
        b.iter(|| {
            let _ = BASE64.encode(&src);
        });
    });
    group.finish();
}

fn bench_base64_dec_5k(c: &mut Criterion) {
    let mut group = c.benchmark_group("Base64Dec/5.3KB");
    group.throughput(Throughput::Bytes(5464));
    let src: Vec<u8> = (0..4096u32).map(|i| (i % 256) as u8).collect();
    let encoded = BASE64.encode(&src);
    group.bench_function("Base64Dec/5.3KB", |b| {
        b.iter(|| {
            let _ = BASE64.decode(&encoded).unwrap();
        });
    });
    group.finish();
}

fn bench_hex_enc_1k(c: &mut Criterion) {
    let mut group = c.benchmark_group("HexEnc/1KB");
    group.throughput(Throughput::Bytes(1024));
    let src: Vec<u8> = (0..1024u32).map(|i| (i % 256) as u8).collect();
    group.bench_function("HexEnc/1KB", |b| {
        b.iter(|| {
            hex::encode(&src)
        });
    });
    group.finish();
}

fn bench_strreplace_10k(c: &mut Criterion) {
    let mut group = c.benchmark_group("StrReplace/10KB");
    group.throughput(Throughput::Elements(10_000));
    let mut s = "x".repeat(100);
    for _ in 0..100 {
        s += "Hello World! This is a test string for replacement. ";
    }
    group.bench_function("StrReplace/10KB", |b| {
        b.iter(|| {
            let _ = s.replace("Hello", "World");
        });
    });
    group.finish();
}

fn bench_json_parse_404b(c: &mut Criterion) {
    let mut group = c.benchmark_group("JSON/Parse/404B");
    group.throughput(Throughput::Bytes(404));
    let s = r#"{"users":[{"id":1,"name":"Alice","email":"alice@example.com","score":95.5,"active":true},{"id":2,"name":"Bob","email":"bob@example.com","score":87.3,"active":false},{"id":3,"name":"Charlie","email":"charlie@example.com","score":92.1,"active":true},{"id":4,"name":"Diana","email":"diana@example.com","score":88.8,"active":true},{"id":5,"name":"Eve","email":"eve@example.com","score":91.0,"active":false}]}"#;
    group.bench_function("JSON/Parse/404B", |b| {
        b.iter(|| {
            let _: serde_json::Value = serde_json::from_str(s).unwrap();
        });
    });
    group.finish();
}

criterion_group!(
    benches,
    bench_inttostr_100k,
    bench_base64_enc_4k,
    bench_base64_dec_5k,
    bench_hex_enc_1k,
    bench_strreplace_10k,
    bench_json_parse_404b
);
criterion_main!(benches);
